### PR TITLE
Purge old schema versions from memory in historian

### PR DESCRIPTION
**Pull request by austenLacy**
## Description

This PR adds the ability for the vttablet to configure purging schema versions _in memory_ used in vreplication schema tracking older than a certain time. The motivation for this change was that for tablets with very large schemas and very frequent DDLs the tablet process would eventually OOM because of storing all of the old schema versions in memory.

> NOTE: This does **_not_** delete the schema version rows from the `_vt.schema_version` table in mysql, but only purges them from the vttablet process memory.

I added a `schema-version-max-age-seconds` vttablet flag that defaults to `0` (no purging) and can be set to purge schema versions older than the set number of seconds. There was an [attempt in the past](https://github.com/vitessio/vitess/pull/8617) to limit the max number of schemas to keep in memory, but was not merged. It didn't account for if you have binlog events with a GTID which could use a schema older than the max number held in memory. Since this approach uses a time based approach the `schema-version-max-age-seconds` can be set to a value greater than your [mysql binlog retention seconds](https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html#sysvar_binlog_expire_logs_seconds) ensuring you don't have any binlogs with a GTID for a schema not in memory.

This work should eventually be replaced by https://github.com/vitessio/vitess/issues/11638.

TODO: 
- [x] write unit tests
- [x] update docs with new vttablet flag

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/8585
- https://github.com/vitessio/vitess/pull/8617

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by mattlord** [comment]
Hi @austenLacy ! 

Thank you for the contribution! I don't see a problem with adding a short term addition that improves the situation until we can do the larger refactor. 

On my first read through though, it seemed that this flag is only consulted at tablet init/startup. I would think you want a way to increase/lower this in the running tablet? If so, we could add it to the `/debug/env` endpoint. 

What do you think?

**Comment by mattlord** [comment]
> @mattlord
> 
> > On my first read through though, it seemed that this flag is only consulted at tablet init/startup. I would think you want a way to increase/lower this in the running tablet?
> 
> I'm not sure if we have a use case for changing the retention dynamically. Do you have any use cases where you'd want the ability to modify the retention at runtime?

I'd defer to you. I didn't think this was an urgent issue that we needed to address before the larger refactor, but for you it is. So I'd use your use case as the use case for now.

> > If so, we could add it to the `/debug/env` endpoint.
> 
> Can you point in the direction of where that code might live?

`go/vt/vttablet/tabletserver/debugenv.go` is the bulk of it.

**Comment by austenLacy** [comment]
> I'd defer to you. I didn't think this was an urgent issue that we needed to address before the larger refactor, but for you it is. So I'd use your use case as the use case for now.

We would likely set this value to the same as our binlog retention seconds on mysql which does not change frequently.

File: go/vt/vttablet/tabletserver/schema/historian.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/historian.go:38]: Don't we want the records time_updated to be _greater_ than the value we pass in? We pass in e.g. now()-1 hour and we only want things that are newer than that, no?

We should test this behavior as well. 

We should also call it out in the PR description.

File: go/vt/vttablet/tabletserver/schema/historian.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/historian.go:175]: What about timezone related issues? I think that we should explicitly use UTC here.

File: go/vt/vttablet/tabletserver/schema/historian_test.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/historian_test.go:215]: It's usually better to  be a bit more specific with the pattern for which you expect a given result. We use the specific query below. Any reason not to here as well?

File: go/vt/vttablet/tabletserver/tabletenv/config.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/tabletenv/config.go:151]: We should use dashes rather than underscores for new flags as we're in the process of moving all flags to dashes.

Does this apply to the query serving schema tracker or only to the vreplication historian? I thought that it was only the latter but the help output makes it seem like the former.

**Review by mattlord — COMMENTED** [review]
Nice work on this @austenLacy ! I only had minor comments. Please let me know what you think.

File: go/vt/vttablet/tabletserver/schema/historian_test.go
Comment (austenLacy) [reply go/vt/vttablet/tabletserver/schema/historian_test.go:215]: I meant to go back and add a more specific regex pattern, but forgot. I'll add that!

File: go/vt/vttablet/tabletserver/tabletenv/config.go
Comment (austenLacy) [reply go/vt/vttablet/tabletserver/tabletenv/config.go:151]: > Does this apply to the query serving schema tracker or only to the vreplication historian? I thought that it was only the latter but the help output makes it seem like the former.

Definitely the vreplication historian. I get the language confused for the different "schema tracking" that is done by both mechanisms. I'll update the help description.

**Comment by austenLacy** [comment]
@mattlord thanks for the review. I've addressed your comments and made the changes.

File: go/vt/vttablet/tabletserver/schema/historian.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/schema/historian.go:38]: We still don't have any tests -- unit or endtoend -- for this part, do we? Something that would have caught the previous issue and will confirm that this works as expected with various inputs. I think we should figure out a way to add a test for it -- with a preference for unit test. I can help with this if you need to bounce some ideas around.

**Review by mattlord — COMMENTED** [review]
I still think we're missing a test, but otherwise it LGTM! Let me know if you could use some help. Reach out to me on the Vitess Slack if so. Thanks!

**Comment by mattlord** [comment]
Also a note that the help text test is failing if you didn't notice already:
```
--- FAIL: TestHelpOutput (0.47s)
    --- FAIL: TestHelpOutput/vttablet (0.07s)
        flags_test.go:117: []: (-want +got)
              (
              	"""
              	... // 262 identical lines
              	      --s3_backup_tls_skip_verify_cert                                   skip the 'certificate is valid' check for SSL connections.
              	      --sanitize_log_messages                                            Remove potentially sensitive information in tablet INFO, WARNING, and ERROR log messages such as query parameters.
            + 	      --schema-version-max-age-seconds int                               max age of schema version records to kept in memory by the vreplication historian
              	      --security_policy string                                           the name of a registered security policy to use for controlling access to URLs - empty means allow all for anyone (built-in policies: deny-all, read-only)
              	      --service_map strings                                              comma separated list of services to enable (or disable if prefixed with '-') Example: grpc-queryservice
              	... // 121 identical lines
              	"""
              )
```

You'll need to update `./go/flags/endtoend/vttablet.txt`

File: go/vt/vttablet/tabletserver/schema/historian.go
Comment (austenLacy) [reply go/vt/vttablet/tabletserver/schema/historian.go:38]: Sounds good. I'll see what I can do and will reach out if I run into any issues.

**Review by mattlord — APPROVED** [review]
LGTM! Thanks again @austenLacy! 

For the record, we sat down and really stepped through what we actually wanted/needed to test for the initial historian query when there's a max age set... and in the end it wasn't really anything more than testing that this golang code worked as it should:
```
time.Now().UTC().Add(time.Duration(-h.schemaMaxAgeSeconds) * time.Second)
```

So given the difficult of adding more for that, we are now instead just confirming that it's the correct query (up to the dynamic value passed in).

File: go/vt/vttablet/tabletserver/schema/historian.go
Comment (rohit-nayak-ps) [inline go/vt/vttablet/tabletserver/schema/historian.go:250]: I think we should do an initial pass to check if there are older schemas that need to be purged and only then create the filtered slice. Otherwise we will be recreating the same slice for a new ddl, if there is no older version to purge.

**Review by rohit-nayak-ps — COMMENTED** [review]
Other than my one comment on optimizing the purge logic, this looks good to me.

File: go/vt/vttablet/tabletserver/schema/historian.go
Comment (austenLacy) [reply go/vt/vttablet/tabletserver/schema/historian.go:250]: Were you thinking something like this? 

```diff
diff --git a/go/vt/vttablet/tabletserver/schema/historian.go b/go/vt/vttablet/tabletserver/schema/historian.go
index 3099408dba..a024b0068c 100644
--- a/go/vt/vttablet/tabletserver/schema/historian.go
+++ b/go/vt/vttablet/tabletserver/schema/historian.go
@@ -248,8 +248,23 @@ func (h *historian) readRow(row []sqltypes.Value) (*trackedSchema, int64, error)
 }
 
 func (h *historian) purgeOldSchemas() {
-       filtered := make([]*trackedSchema, 0)
        maxAgeDuration := time.Duration(h.schemaMaxAgeSeconds) * time.Second
+       shouldPurge := false
+
+       // check if we have any schemas we need to purge and only create the filtered
+       // slice if necessary
+       for _, s := range h.schemas {
+               if time.Since(time.Unix(s.timeUpdated, 0)) > maxAgeDuration {
+                       shouldPurge = true
+                       break
+               }
+       }
+
+       if !shouldPurge {
+               return
+       }
+
+       filtered := make([]*trackedSchema, 0)
        for _, s := range h.schemas {
                if time.Since(time.Unix(s.timeUpdated, 0)) < maxAgeDuration {
                        filtered = append(filtered, s)
```

Comment (austenLacy) [reply go/vt/vttablet/tabletserver/schema/historian.go:250]: Done in https://github.com/vitessio/vitess/pull/13056/commits/2a787baf81885e6082caf6b912bc056570a5a5c4

**Review by rohit-nayak-ps — APPROVED** [review]

**Comment by mattlord** [comment]
@austenLacy can you please create a docs PR in the website repo and add a link to that in the description? Ping me on slack if you need any help with that. Once we at least have that linked (doesn't necessarily need to be 100% done) then we can get this merged! :shipit:

**Comment by austenLacy** [comment]
@mattlord docs PR: https://github.com/vitessio/website/pull/1475

**Pull request merged**
